modules/collect/src/main/java/com/opengamma/collect/id/StandardLink.java

```diff
@@ -0,0 +1,533 @@
+ * <p>
+ * This is the standard implementation of {@link Link}.
+ * It is suitable for all link targets except those that have generic parameters.
+ *
+ * @param <T> the type of the target
+ */
+@BeanDefinition(builderScope = "private")
+public final class StandardLink<T extends IdentifiableBean>
+    implements Link<T>, Resolvable<StandardLink<T>>, StandardIdentifiable, ImmutableBean, Serializable {
+
+  /**
+   * The primary standard identifier of the target.
+   * <p>
+   * The standard identifier is used to identify the target.
+   * It will typically be an identifier in an external data system.
+   * <p>
+   * This is used when the link is in the resolvable state.
+   */
+  @PropertyDefinition(get = "field")
+  private final StandardId standardId;
+  /**
+   * The type of the link target.
+   * <p>
+   * This is used when the link is in the resolvable state.
+   */
+  @PropertyDefinition(get = "field")
+  private final Class<T> targetType;
+  /**
+   * The embedded link target.
+   * <p>
+   * This is used when the link is in the resolved state.
+   */
+  @PropertyDefinition(get = "field")
+  private final T target;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Create a resolvable link for the specified identifier and type.
+   * <p>
+   * The returned link will be in the resolvable state, containing the identifier and type.
+   * When the {@link #resolve(LinkResolver)} method is called, the resolver will be used
+   * to find the target, for example from a database.
+   * <p>
+   * The type is expressed as a standard {@link Class} object.
+   *
+   * @param identifier  the primary identifier of the target
+   * @param targetType  the type of the target
+   * @return a new resolvable link
+   */
+  public static <R extends IdentifiableBean> StandardLink<R> resolvable(StandardId identifier, Class<R> targetType) {
+    return new StandardLink<>(identifier, targetType);
+  }
+
+  /**
+   * Create a link with the link target embedded directly.
+   * <p>
+   * The returned link will be in the resolved state, directly embedding the target object.
+   * When the {@link #resolve(LinkResolver)} method is called, the embedded target will be
+   * returned without needing to use the resolver.
+   *
+   * @param target  the link target
+   * @return a new resolved link
+   */
+  public static <R extends IdentifiableBean> StandardLink<R> resolved(R target) {
+    return new StandardLink<>(target);
+  }
+
+  //-------------------------------------------------------------------------
+  // constructor for resolvable state
+  private StandardLink(StandardId identifier, Class<T> targetType) {
+    this.standardId = ArgChecker.notNull(identifier, "identifier");
+    this.targetType = ArgChecker.notNull(targetType, "targetType");
+    this.target = null;
+  }
+
+  // constructor for resolved state
+  private StandardLink(T target) {
+    this.standardId = null;
```

Comment:
Design ensures that the serialized form is simple and correctly represents the "either" nature of the object.